### PR TITLE
Add ability to specify hardware target and jtag frequency when progra…

### PR DESCRIPTION
**Pull request by benreynwar**
…mming with vivado.

### CODE REVIEW

File: edalize/templates/vivado/vivado-makefile.j2 [hunk 1]
```diff
@@ -8,5 +8,14 @@ $(NAME).bit:  $(NAME)_run.tcl $(NAME).xpr
 $(NAME).xpr: $(NAME).tcl
 	vivado -mode batch -source $<
 
+ifdef HWTARGET
+program: $(NAME).bit $(NAME)_pgm.tcl
+	vivado -mode batch -source $(NAME)_pgm.tcl -tclargs $(HWTARGET)
+else
+program: $(NAME).bit $(NAME)_pgm.tcl
+	vivado -mode batch -source $(NAME)_pgm.tcl
+endif
```
Comment (imphil) [inline edalize/templates/vivado/vivado-makefile.j2]: Why not just set `HWTARGET` as an environment variable (if it is passed as `make` variable, convert it to an environment variable here). You can then pick it up from the TCL script and don't need to rely on fragile argv parsing.

File: edalize/templates/vivado/vivado-program.tcl.j2 [hunk 2]
```diff
@@ -1,33 +1,44 @@
...
 
 set found 0
 
-foreach { hw_target } [get_hw_targets] {
+if {$hw_target == ""} {
+    set hw_targets [get_hw_targets]
+} else {
+    set hw_targets [get_hw_targets $hw_target]
+}
+
```
Comment (olofk) [inline edalize/templates/vivado/vivado-program.tcl.j2]: I think we can simplify this. If $hw_target is empty then I believe [get_hw_targets $hw_target] will just evaluate to [get_hw_targets].

**Comment by olofk** [comment]
Sorry for the delay. I added a comment on TCL simplification.

Another more general question is if hw_target is the best way to distinguish a certain target. I don't have enough experience with this, but they seem globally unique. My two Digilent Nexys A7 boards report
`localhost:3121/xilinx_tcf/Digilent/210292AA7815A` and `localhost:3121/xilinx_tcf/Digilent/210292AA7829A` so I guess it should be fine

I'm fine with this patch once it has been cleaned up a bit, but I really really hope someone will eventually start a separate abstraction library for FPGA programming devices.

**Comment by benreynwar** [comment]
Just got bitten by the fact that I forgot about this and never worked to get it merged in :).
Turns out that if your programming step is silently failing then debugging the design running on the FPGA becomes really confusing!
I still want to test this a little more so am marking it WIP.

**Comment by olofk** [comment]
Cool. Thanks for working on it. Let me know when it's in a reviewable state

File: edalize/templates/vivado/vivado-program.tcl.j2 (hunk 2 above)
Comment (benreynwar) [reply edalize/templates/vivado/vivado-program.tcl.j2]: Just checked this and it didn't work.  Only took me 9 months :).

**Comment by benreynwar** [comment]
This is ready for a review when you get a chance.  Thanks.

**Comment by olofk** [comment]
Thanks! Squashed and merged. I took the liberty to slightly reword the commit message. Hope that's fine

**Comment by benreynwar** [comment]
Thanks @olofk.  Yes, feel free to modify anything of mine whenever you want :).  I'll try not to break the tests next time too!

**Pull request merged**